### ISSUE TITLE
Proposed rule: warn when calling Guid parameterless ctor

### ISSUE THREAD

**Issue by GrabYourPitchforks** [dotnet/runtime#43973]
Based on a conversation at https://twitter.com/twwilliams/status/1190312876927401984?s=21.

It's easy for developers to inadvertently write `new Guid()` in their code when they likely meant to write `Guid.NewGuid`. Since `Guid` is a value type, the ctor zero-inits the returned instance.

This proposal is for an analyzer to detect calls to this ctor and to produce a warning: direct the author to call `NewGuid()` instead if they intended to generate a new `Guid`, or explicitly write `default(Guid)` or `Guid.Empty` if they intended to zero-init.

Edit: Modified based on a suggestion from https://twitter.com/Jjagg_dev/status/1190799663441555456.

**Comment by stephentoub** [comment]
> Let's move it to the runtime repo and prepare it for API review.

FWIW, note that dotnet/runtime has been using https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1129.md for this purpose for a while now.

**Comment by buyaa-n** [comment]
> Suggestion:
> 
> 1. Implement this rule as _disabled by default_.
> 2. When the rule is enabled by the user via a ruleset/editorconfig entry, the rule fires on all value types like @stephentoub suggested. We can special case the diagnostic message for `Guid`.
> 3. Make the rule configurable so users can restrict the type names on which it fires. For example, introduce a new .editorconfig option `applicable_symbol_names` similar to [disallowed-symbol-names](https://github.com/dotnet/roslyn-analyzers/blob/master/docs/Analyzer%20Configuration.md#disallowed-symbol-names).

[Triage]: Seems we are agreeing to have a rule for all value types (similar to [SA1129: Do not use default value type constructor](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1129.md)) and special case the `Guid`, questions:

1. How to special case the `Guid`? Only the message? Or fixer also should suggest 2 options like `Guid.NewGuid` and `default(Guid)`? 
2. Is `Guid` is the only type that needs special casing?

**Comment by stephentoub** [comment]
FWIW, I still think this is either a) something we should leave for stylecop to handle (it already has a perfectly good rule for this, that we're using) or b) or should be something the IDE/Roslyn drives along with all of the other style rules. Our focus for analyzers is on correctness/performance/etc, and shouldn't be style, plus having this style one-off is inconsistent from the rest of the style experiences in VS. There's an argument to be made that while this is style, it highlights potential misuse and thus is borderline correctness, but that's true of a lot of style, actually. On top of which, it's even more of a style decision now that structs support parameterless public ctors, so folks could easily want the consistency of always using new.

**Comment by StingyJack** [comment]
I dont think this is a style issue because the problem is not `= default` vs `new Guid()`.  Its that many programmers make the mistake of thinking that `new Guid()` is going to, you know,... make a new globally unique identifier. The way that it initializes to 0's (a value that isn't globally unique) is not intuitive  and not consistent with other non-numerical types (strings and booleans dont initialize to "0").

I'm sure there was/is probably a good reason to do this 0's initialization for Guids, Since changing the framework to make `new Guid()` behave like `NewGuid()` is not realistic at this point, this rule is needed to address a common **incorrect usage** scenario that has been getting made repeatedly for the last 20 years. 

In my experience, most development teams dont have any kind of code analysis outside of what VS and MSB provides as warnings and errors out of the box. StyleCop is a luxury most teams dont have (or cant for various reasons).

**Comment by GrabYourPitchforks** [comment]
I agree with @StingyJack's comment [above](https://github.com/dotnet/runtime/issues/43973#issuecomment-1089290922), which also reflects what I mentioned [earlier](https://github.com/dotnet/runtime/issues/43973#issuecomment-549579561). We know that `new Guid()` is particularly unintuitive and troublesome. It has [appeared](https://stackoverflow.com/questions/7972150/guid-is-all-0s-zeros) [multiple](https://stackoverflow.com/questions/11938151/guid-newguid-vs-new-guid) [times](https://stackoverflow.com/questions/904021/how-to-use-guids-in-c) on StackOverflow as something that trips people up.

If somebody wants to create an analyzer / style rule / whatever that flags all parameterless struct creation, please feel free to do that. But I think the `new Guid()` pattern deserves its own rule because of how problematic it is. We have the capability to solve this specific pain point. And I would prefer to untie the `Guid`-specific issue from the more general conversation about structs, since one is about correctness and the other is largely about style.

**Comment by StingyJack** [comment]
> I still do not see why Guid is so special here.

Because "Declare and instantiate a new Globally Unique Identifier" (aka `var g = new Guid();`)  incorrectly sets the expectation that the variable will have a Globally Unique Identifier in it, and not a non-unique 0 equivalent.

It's inconsistent with all the other basic types people will use in programs. 
``` c#
var s = new string(); //expected to be an empty string
int i;  //expected to be the intuitive default of 0
var i = new int(); // just like another value type - new Guid()  - right?
var d = new DateTime(....) has a bunch of overloads, and each of them give you the expected result. 
```

> because `new()` for structs is a general pit-of-failure

In the last 15 years I've worked at small and large companies in several different industries, and have probably read through the source code of at least a thousand project's, and I can count on less than one hand how many times I can remember seeing a user defined struct. If we were to ask the average developer what a struct was they could probably recite the textbook definition but very few would have anything more to say about it (a few of us that read [Framework Design Guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct) would probably cite the advice from that, which is summarized as "use classes"). I dont think I can name 5 commonly used structs in .net, The point I'm trying to make is that we dont see the pitfall  except for `new Guid()`

OTOH, you probably see them all the time and are thus more familiar with the pitfalls.

**Comment by buyaa-n** [comment]
> I still think this is either a) something we should leave for stylecop to handle (it already has a perfectly good rule for this, that we're using)

Got it, lets leave the general rule for `default(Struct)` and `new Struct()` for the stylecop to handle

> I still do not see why Guid is so special here. The previously mentioned rules exist largely because `new()` for structs is a general pit-of-failure, with folks expecting it is invoking a constructor (and until recently it not even being possible to define such a ctor in C#), and the "style" of it emerged as a way to prevent those failures,

By my understanding it is rare that a developer mistakenly use `new Guid()` instead of `default(Guid)`, especially in normal application development a developer most likely wanted to generate a **new unique Guid** and could mistakenly use `new Guid()` instead of `Guid.NewGuid()`. The `StackOverflow` findings @GrabYourPitchforks shared above shows that it is quite common mistake: [hit 1](https://stackoverflow.com/questions/7972150/guid-is-all-0s-zeros), [hit 2](https://stackoverflow.com/questions/11938151/guid-newguid-vs-new-guid), [hit 3](https://stackoverflow.com/questions/904021/how-to-use-guids-in-c)

**Comment by CurtHagenlocher** [comment]
This rule would have saved us from a privacy incident :(.

**Comment by tannergooding** [comment]
Going to close this. There's not been enough consensus to drive it forward and its just polluting the backlog as is. At best these seems to be an opt-in analyzer that won't help unless users know to enable it. StyleCop has an analyzer that covers this already as well, so users aren't blocked.

This is also reasonably something that should be caught by trivial testing. While it isn't your job to validate that `NewGuid` actually produces random values, it is expected that your code do things like validate simple logic actually touches two distinct entries, for example, as that could go wrong in many ways.
